src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+    Af = assemble_matrix(fem_params.Pf, fem_params.Qf) do u, v
+        ∫(a_f(r, u, v))fem_params.dΩ_d + ∫(v * u)fem_params.dΩ_d
+    end
+    wvec = Af' \ dgdpf
+    wh = FEFunction(fem_params.Pf, wvec)
+    l_temp(dp) = ∫(wh * dp)fem_params.dΩ_d
+    return assemble_vector(l_temp, fem_params.P)
+end
+
+# Finally, we pack up everything and rewrite it as a compact form. We also record every optimization step values.
+# 
+
+function gf_p(p0::Vector; r, β, η, phys_params, fem_params)
+    pf_vec = pf_p0(p0; r, fem_params)
+    gf_pf(pf_vec; β, η, phys_params, fem_params)
+end
+
+function gf_p(p0::Vector, grad::Vector; r, β, η, phys_params, fem_params)
+    if length(grad) > 0
+        dgdp, = Zygote.gradient(p -> gf_p(p; r, β, η, phys_params, fem_params), p0)
+        grad[:] = dgdp
+    end
+    gvalue = gf_p(p0::Vector; r, β, η, phys_params, fem_params)
+    open("gvalue.txt", "a") do io
+        write(io, "$gvalue \n")
+    end
+    gvalue
+end
+
+# Using the following codes, we can check if we can get the derivatives correctly from the adjoint method by comparing it with the finite difference results.
+# 
+
+p0 = rand(fem_params.np)
+δp = rand(fem_params.np)*1e-8
+grad = zeros(fem_params.np)
+
+g0 = gf_p(p0, grad; r, β, η, phys_params, fem_params)
+g1 = gf_p(p0+δp, []; r, β, η, phys_params, fem_params)
+g1-g0, grad'*δp
+
+# ## Optimization with  NLop
+# 
+# Now we use NLopt.jl package to implement the MMA algorithm for optimization. Note that we start with $\beta=8$ and then gradually increase it to $\beta=32$ in consistant with Ref. [6].
+# 
+
+using NLopt, DelimitedFiles
+
+function gf_p_optimize(p_init, r, β, η, TOL = 1e-4, MAX_ITER = 500; phys_params, fem_params)
+    ##################### Optimize #################
+    opt = Opt(:LD_MMA, fem_params.np)
+    lb = zeros(fem_params.np)
+    ub = ones(fem_params.np)
+    opt.lower_bounds = lb
+    opt.upper_bounds = ub
+    opt.ftol_rel = TOL
+    opt.maxeval = MAX_ITER
+    opt.max_objective = (p0, grad) -> gf_p(p0, grad; r, β, η, phys_params, fem_params)
+    if (length(p_init)==0)
+        p_initial_guess = readdlm("p_opt_value.txt", Float64)
+        p_initial_guess = p_initial_guess[:]
+    else
+        p_initial_guess = p_init[:]
+    end
+
+    (g_opt, p_opt, ret) = optimize(opt, p_initial_guess)
+    @show numevals = opt.numevals # the number of function evaluations
+    
+    return g_opt, p_opt
+end
+
+p_init = ones(fem_params.np) * 0.4   # Initial guess
+β_list = [8.0, 16.0, 32.0]
+
+g_opt = 0
+for bi = 1 : 3
+    β = β_list[bi]
+    if bi == 1
+        g_opt, p_opt = gf_p_optimize(p_init, r, β, η, 1e-8, 100; phys_params, fem_params)
+    
+    else
+        g_opt, p_opt = gf_p_optimize([], r, β, η, 1e-8, 100; phys_params, fem_params)
+    end
+    open("p_opt_value.txt", "w") do iop
+        for i = 1 : length(p_opt)
+            write(iop, "$(p_opt[i]) \n")
+        end
+    end
+    open("g_opt_value.txt", "a") do io
+        write(io, "$g_opt \n")
+    end
```

Comment:
Same here, I would just `@show g_opt`, or save into an array if you want to plot later in the script.

I get that these are long-running calculations and you often want to do checkpointing and logging of intermediate steps, but the tutorial doesn't seem like the place for this.   

(For example, if you just want to log the values, then when you run the `.jl` file you could simply pipe it to a file with `julia TopOptEMFocus.jl > log.txt`.)